Source/Core/Core/AchievementManager.cpp

```diff
@@ -41,15 +43,50 @@ bool AchievementManager::IsLoggedIn() const
   return m_login_data.response.succeeded;
 }
 
+void AchievementManager::LoadGameByFilenameAsync(const std::string& iso_path,
+                                                 const ResponseCallback& callback)
+{
+  rc_hash_generate_from_file(game_hash, RC_CONSOLE_GAMECUBE, iso_path.c_str());
+  ResponseCallback fetch_game_data_callback = [this, callback](ResponseType r_type) {
+    m_is_game_loaded = (r_type == ResponseType::SUCCESS);
+    callback(r_type);
+  };
+  ResponseCallback start_session_callback = [this, fetch_game_data_callback](ResponseType r_type) {
+    if (r_type == ResponseType::SUCCESS)
+      m_queue.EmplaceItem(
+          [this, fetch_game_data_callback] { fetch_game_data_callback(FetchGameData()); });
+  };
+  ResponseCallback resolve_hash_callback = [this, start_session_callback](ResponseType r_type) {
+    if (r_type == ResponseType::SUCCESS)
+      m_queue.EmplaceItem(
+          [this, start_session_callback] { start_session_callback(StartRASession()); });
+  };
+  m_queue.EmplaceItem([this, resolve_hash_callback] { resolve_hash_callback(ResolveHash()); });
+}
```

Comment:
The way you've written this function is extremely confusing to me. Is there anything I'm missing, or couldn't this just be a sequence of function calls, ie something like

```
void AchievementManager::LoadGameByFilenameAsync(const std::string& iso_path,
                                                 const ResponseCallback& callback)
{
  rc_hash_generate_from_file(game_hash, RC_CONSOLE_GAMECUBE, iso_path.c_str());
  m_queue.EmplaceItem([this, callback] {
    const auto resolve_hash_response = ResolveHash();
    if (resolve_hash_response != ResponseType::SUCCESS)
    {
      callback(resolve_hash_response);
      return;
    }

    const auto start_session_response = StartRASession();
    if (start_session_response != ResponseType::SUCCESS)
    {
      callback(start_session_response);
      return;
    }

    const auto fetch_game_data_response = FetchGameData();
    m_is_game_loaded = fetch_game_data_response == ResponseType::SUCCESS;
    callback(fetch_game_data_response);
  });
}
```

Reply:
For some reason it never once occurred to me that I could just emplace multiple API calls in the same work queue item. This _does_ make a lot more sense.